### ISSUE TITLE
gu_asio_test.cpp failed: what():  Failed to connect UDP socket: set_option: No such device: 19 (No such device)

### ISSUE THREAD

**Issue by Whissi** [codership/galera#595]
On Gentoo Linux, we see the following failure:
```
Running suite(s): gu::asio
terminate called after throwing an instance of 'gu::Exception'
  what():  Failed to connect UDP socket: set_option: No such device: 19 (No such device)
         at galerautils/src/gu_asio_datagram.cpp:connect():188
98%: Checks: 80, Failures: 0, Errors: 1
galerautils/tests/gu_asio_test.cpp:1816:E:test_datagram_connect_multicast:test_datagram_connect_multicast:0: (after this point) Received signal 6 (Aborted)
Running suite(s): gu::DeqMap
100%: Checks: 11, Failures: 0, Errors: 0
Total tests failed: 1
scons: *** [galerautils/tests/gu_tests++.passed] Error 1
scons: building terminated because of errors.
```

**Comment by temeo** [comment]
Does this patch fix the problem?
```
diff --git a/galerautils/tests/gu_asio_test.cpp b/galerautils/tests/gu_asio_test.cpp
index bf870a65..e82cadfc 100644
--- a/galerautils/tests/gu_asio_test.cpp
+++ b/galerautils/tests/gu_asio_test.cpp
@@ -2314,7 +2314,7 @@ Suite* gu_asio_suite()
     tcase_add_test(tc, test_datagram_send_to_and_async_read);
     suite_add_tcase(s, tc);
 
-#if defined(__FreeBSD__)
+#if !defined(GALERA_ASIO_TEST_MULTICAST)
     /* fails on FreeBSD with EADDRNOTAVAIL, disable temporarily */
     (void)test_datagram_send_to_and_async_read_multicast;
 #else
```

**Comment by Whissi** [comment]
No. But if I understand you correctly, you want to make this test skipped (because GALERA_ASIO_TEST_MULTICAST isn't defined). Wouldn't we need to move this if clause up, i.e. to start before

https://github.com/codership/galera/blob/d54ebf885e302380d1724761220f970a32327fe6/galerautils/tests/gu_asio_test.cpp#L2278

**Comment by temeo** [comment]
Yes, that's correct, all the multicast tests should be disabled by default. This should do it:
```
diff --git a/galerautils/tests/gu_asio_test.cpp b/galerautils/tests/gu_asio_test.cpp
index bf870a65..a76022c3 100644
--- a/galerautils/tests/gu_asio_test.cpp
+++ b/galerautils/tests/gu_asio_test.cpp
@@ -2302,6 +2302,11 @@ Suite* gu_asio_suite()
     tcase_add_test(tc, test_datagram_open_connect);
     suite_add_tcase(s, tc);
 
+    tc = tcase_create("test_datagram_send_to_and_async_read");
+    tcase_add_test(tc, test_datagram_send_to_and_async_read);
+    suite_add_tcase(s, tc);
+
+#if defined(GALERA_ASIO_TEST_MULTICAST)
     tc = tcase_create("test_datagram_connect_multicast");
     tcase_add_test(tc, test_datagram_connect_multicast);
     suite_add_tcase(s, tc);
@@ -2310,22 +2315,19 @@ Suite* gu_asio_suite()
     tcase_add_test(tc, test_datagram_connect_multicast_local_if);
     suite_add_tcase(s, tc);
 
-    tc = tcase_create("test_datagram_send_to_and_async_read");
-    tcase_add_test(tc, test_datagram_send_to_and_async_read);
-    suite_add_tcase(s, tc);
-
-#if defined(__FreeBSD__)
-    /* fails on FreeBSD with EADDRNOTAVAIL, disable temporarily */
-    (void)test_datagram_send_to_and_async_read_multicast;
-#else
     tc = tcase_create("test_datagram_send_to_and_async_read_multicast");
     tcase_add_test(tc, test_datagram_send_to_and_async_read_multicast);
     suite_add_tcase(s, tc);
-#endif /* FreeBSD */
 
     tc = tcase_create("test_datagram_write_multicast");
     tcase_add_test(tc, test_datagram_write_multicast);
     suite_add_tcase(s, tc);
+#else
+    (void)test_datagram_connect_multicast;
+    (void)test_datagram_connect_multicast_local_if;
+    (void)test_datagram_send_to_and_async_read_multicast;
+    (void)test_datagram_write_multicast;
+#endif /* GALERA_ASIO_TEST_MULTICAST */
 
     //
     // Steady timer
```

**Comment by temeo** [comment]
Should be fixed in 4.9.
